mitodl/mit-learn
# Shanbady qdrant integration for vector based similar resources endpoint

### What are the relevant tickets?

Closes https://github.com/mitodl/hq/issues/5865

### Description (What does it do?)
This PR introduces the following
- a new endpoint to show similar resources using vector embeddings via [QDrant](https://qdrant.tech/)
- a "generate_embeddings" management command to embed learning resources and contentfiles

### How can this be tested?
1. Checkout this branch
2. You will need to rebuild your images to install dependencies. (`docker compose down` `docker compose build` )
3. load up learning resources via management command
4. run the embedding management command `./manage.py generate_embeddings --all --skip-content-files` (skip contentfiles for now - more on this in the additional context section)
5. once embeddings have been generated, visit the [qdrant dashboard](http://localhost:6333/dashboard) and see that the resources are populated.
6. find a course id and visit the new similar resources endpoint http://open.odl.local:8063/api/v1/learning_resources/**<<resource_id>>**/vector_similar/ and validate the results

### Additional Context
1. The results do not currently at contentfiles (just learning resources). Contentfiles are also stored in a separate collection (unlike in opensearch where they sit with the index of the associated learning resource). Once this goes out and we see this works get a sense of performance, it should be easy to follow up with the addition of contentfile data . Either keep it in a separate collection and make the relation to its parent resource in code or embed learning resources with contentfile vectors in a single collection.
2. The "generate_embeddings" management command is currently not based into the celery crontab - Once this is deployed and manually validated, I'll follow up with a pr that adds this into the celery schedule.
3. Contentfiles (atleast locally) take a long time but i dont think much longer than opensearch indexing. 
4. I was running into some uwsgi exception locally which appears to be resolved by setting the harakiri timeout in uwsgi.ini. Not sure if we expect this to happen when deployed but we can set it there if we need do.
